### ISSUE TITLE
Some questions about the difference with  author's code

### ISSUE THREAD

**Issue by EvanIWNL** [mirzaim/VisionGNN#2]
1. Is there no implementation with a pyramid shape?
2. I noticed the author's code uses convolution for upsampling(or patch embedding), why doesn't your code include it?

And many models achieve accuracy above 80% on this dataset. Since you have dedicated research on this, I'd like to ask for your insights on the performance of this particular model.

Thanks XD

**Comment by morteza-im** [comment]
Hi @EvanIWNL,  

Thanks for the great questions!  

1. No, I didn’t implement the pyramid shape. This was actually a course project where we were tasked with choosing a paper and investigating its re-implementability, even if only partially. I selected this paper for that purpose.  

2. After completing my implementation, I discovered that the original authors used convolutional layers for feature extraction. However, this wasn’t explicitly mentioned in the paper—or at least I couldn’t find it. The goal of the project was to try to reimplement the paper based solely on the information provided in the paper itself, without referencing the official or other implementations. I felt there should be some feature extractor in the first layer, so I used a linear feature extractor, which worked fine. However, it’s possible that CNNs might be more efficient for this task.  

Here’s the feature extractor I implemented:  

```python
self.patch_embedding = nn.Sequential(
    nn.Linear(in_features, out_feature // 2),
    nn.BatchNorm1d(out_feature // 2),
    nn.GELU(),
    nn.Linear(out_feature // 2, out_feature // 4),
    nn.BatchNorm1d(out_feature // 4),
    nn.GELU(),
    nn.Linear(out_feature // 4, out_feature // 8),
    nn.BatchNorm1d(out_feature // 8),
    nn.GELU(),
    nn.Linear(out_feature // 8, out_feature // 4),
    nn.BatchNorm1d(out_feature // 4),
    nn.GELU(),
    nn.Linear(out_feature // 4, out_feature // 2),
    nn.BatchNorm1d(out_feature // 2),
    nn.GELU(),
    nn.Linear(out_feature // 2, out_feature),
    nn.BatchNorm1d(out_feature)
)
```

**Comment by EvanIWNL** [comment]
> Hi @EvanIWNL,
> 
> Thanks for the great questions!
> 
> 1. No, I didn’t implement the pyramid shape. This was actually a course project where we were tasked with choosing a paper and investigating its re-implementability, even if only partially. I selected this paper for that purpose.
> 2. After completing my implementation, I discovered that the original authors used convolutional layers for feature extraction. However, this wasn’t explicitly mentioned in the paper—or at least I couldn’t find it. The goal of the project was to try to reimplement the paper based solely on the information provided in the paper itself, without referencing the official or other implementations. I felt there should be some feature extractor in the first layer, so I used a linear feature extractor, which worked fine. However, it’s possible that CNNs might be more efficient for this task.
> 
> Here’s the feature extractor I implemented:
> 
> ```python
> self.patch_embedding = nn.Sequential(
>     nn.Linear(in_features, out_feature // 2),
>     nn.BatchNorm1d(out_feature // 2),
>     nn.GELU(),
>     nn.Linear(out_feature // 2, out_feature // 4),
>     nn.BatchNorm1d(out_feature // 4),
>     nn.GELU(),
>     nn.Linear(out_feature // 4, out_feature // 8),
>     nn.BatchNorm1d(out_feature // 8),
>     nn.GELU(),
>     nn.Linear(out_feature // 8, out_feature // 4),
>     nn.BatchNorm1d(out_feature // 4),
>     nn.GELU(),
>     nn.Linear(out_feature // 4, out_feature // 2),
>     nn.BatchNorm1d(out_feature // 2),
>     nn.GELU(),
>     nn.Linear(out_feature // 2, out_feature),
>     nn.BatchNorm1d(out_feature)
> )
> ```

**Comment by EvanIWNL** [comment]
@mirzaim Actually I also found much unmentioned design in their code. Your code is much more concise and easier to understand. It helps a lot.
